TeamCode/src/main/java/org/firstinspires/ftc/teamcode/pedroPathing/localization/localizers/AprilTagLocalizer

```diff
@@ -0,0 +1,317 @@
+                .build();
+
+
+        // TODO: update this
+        cameraOffset = new Pose(0,0);
+
+        setStartPose(setStartPose);
+        totalHeading = 0;
+        previousHeading = startPose.getHeading();
+    }
+    public Pose getVectorBasedOnTags() {
+        ArrayList<AprilTagDetection> detections = aprilTag.getDetections();
+        if (detections.isEmpty()) {
+            return null;
+        } else {
+            tagDetectTime = aprilTag.getDetections().get(0).frameAcquisitionNanoTime;
+
+            Vector2d posVector2d = aprilTag.getDetections().stream() // get the tag detections as a Java stream
+                    // convert them to Vector2d positions using getFCPosition
+                    .map(detection -> getFCPosition(detection, backupLocalizer.getPose().getHeading(), cameraOffset))
+                    // add them together
+                    .reduce(new Vector2d(0, 0), Vector2d::plus)
+                    // divide by the number of tags to get the average position
+                    .div(aprilTag.getDetections().size());
+
+            return new Pose(posVector2d.x, posVector2d.y);
+        }
+    }
+
+
+    /**
+     * This returns the current pose estimate.
+     *
+     * @return returns the current pose estimate as a Pose
+     */
+    @Override
+    public Pose getPose() {
+        Pose pose = getVectorBasedOnTags();
+
+        if (pose != null) {
+            double x = pose.getX();
+            double y = pose.getY();
+
+            if (XAxisInverted) x = x * -1;
+            if (YAxisInverted) y = y * -1;
+
+            // update secondary localizer
+            backupLocalizer.setPose(new Pose(x, y));
+
+            return MathFunctions.addPoses(startPose, new Pose(x, y, backupLocalizer.getTotalHeading()));
+        } else {
+            // in case the robot couldn't detect any tags
+            return backupLocalizer.getPose();
+        }
+    }
+
+    // Thanks to j5155 and Michael from team 14343 for creating the vast majority of FCPosition(); and getVectorBasedOnTags();
+
+    public Vector2d getFCPosition(AprilTagDetection detection, double botheading, Pose cameraOffset) {
+        // get coordinates of the robot in RC coordinates
+        // ensure offsets are RC
+        double x = detection.ftcPose.x - cameraOffset.getX();
+        double y = detection.ftcPose.y - cameraOffset.getY();
+
+        // invert heading to correct properly
+        botheading = -botheading;
+
+        // rotate RC coordinates to be field-centric
+        double x2 = x*Math.cos(botheading)+y*Math.sin(botheading);
+        double y2 = x*-Math.sin(botheading)+y*Math.cos(botheading);
+        // add FC coordinates to apriltag position
+        // tags is just the CS apriltag library
+        VectorF tagpose = getCenterStageTagLibrary().lookupTag(detection.id).fieldPosition;
+
+
+        // todo: this will need to be changed for next season (use tag heading to automate??)
```

Comment:
I never changed this for this season, so this assumes you're still in Centerstage tags. You'll need to update it.

Reply:
Yeah I'm aware, I haven't had a chance to finish this, but I will try to get it done eventually.